<a href="https://colab.research.google.com/github/String-ru/AIFFEL_quest_cr/blob/main/Datathon/1st_datathon_0220_4PM_1st_assumption.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#1st Assumption

#특정 배우는 특정 장르에서 더 높은 선호도를 보일까?

(장르에 대한 배우 개인의 선호도X)

[여러 장르에 출연하는 배우에 대한 시청자의 선호도O]

**특정 배우의 특정 장르에 대한 출연을 선호하는 경향이 시청자들 사이에 존재하는가?**

#개요

1. 배우별 장르 출연 횟수 세야 함

2. 장르별 시청자 선호도 파악 (imdb data set)

3. 배우와 장르의 관계파악 (특정 배우가 주로 출연하는 장르에서 시청자의 선호도가 높은지 여부 파악)

4. 그 관계 -> 시각화 & 통계적 분석

# 목차

# 1. Data 불러오기

Google Drive 마운트 및 OTT 데이터셋 로드, 컬럼 정리, 플랫폼 정보 추가 완료. ✅

# 2. 모든 dataset에서 column 동일한지 확인

모든 OTT 데이터셋의 컬럼이 동일한지 검증 완료. ✅

# 1. Data 불러오기

Google Drive 마운트 및 OTT 데이터셋 로드, 컬럼 정리, 플랫폼 정보 추가 완료. ✅

In [13]:
# 1️⃣ Google Drive 마운트
from google.colab import drive
drive.mount('/content/drive')

# 2️⃣ 필요한 라이브러리 불러오기
import os
import pandas as pd

# 3️⃣ 데이터 폴더 경로 설정
data_folder = "/content/drive/MyDrive/1st_datathon"  # 폴더 경로

# 4️⃣ 폴더 내의 파일 목록 확인
file_list = os.listdir(data_folder)
print("📂 데이터 폴더 내 파일 목록:", file_list)

# 5️⃣ CSV 파일 경로 설정
disney_file = os.path.join(data_folder, "disney_plus_titles.csv")
amazon_file = os.path.join(data_folder, "amazon_prime_titles.csv")
netflix_file = os.path.join(data_folder, "netflix_titles.csv")

# 6️⃣ pandas로 CSV 파일 로드
disney_df = pd.read_csv(disney_file)
amazon_df = pd.read_csv(amazon_file)
netflix_df = pd.read_csv(netflix_file)

# 7️⃣ 컬럼 이름을 소문자로 통일
netflix_df.columns = netflix_df.columns.str.lower()
amazon_df.columns = amazon_df.columns.str.lower()
disney_df.columns = disney_df.columns.str.lower()

# 8️⃣ 공통 컬럼만 추출 (필요한 컬럼만 남기는 것으로 활용해도 OK)
common_cols = ['show_id', 'type', 'title', 'director', 'cast', 'release_year', 'date_added','listed_in','duration']
netflix_df = netflix_df[common_cols]
amazon_df = amazon_df[common_cols]
disney_df = disney_df[common_cols]

# 9️⃣ 출처 플랫폼 컬럼 추가
netflix_df['platform'] = 'Netflix'
amazon_df['platform'] = 'Amazon'
disney_df['platform'] = 'Disney'

# 🔟 데이터 크기 확인
print("✅ Disney+ 데이터:", disney_df.shape)
print("✅ Amazon Prime 데이터:", amazon_df.shape)
print("✅ Netflix 데이터:", netflix_df.shape)

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
📂 데이터 폴더 내 파일 목록: ['disney_plus_titles.csv', 'amazon_prime_titles.csv', 'netflix_titles.csv']
✅ Disney+ 데이터: (1450, 10)
✅ Amazon Prime 데이터: (9668, 10)
✅ Netflix 데이터: (8807, 10)


<ipython-input-13-10468c4c6dc2>:40: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  disney_df['platform'] = 'Disney'


=> 각각 1450 / 9668 / 8807개의 항목(영화나 TV쇼 etc.)에 대한 12개의 정보(column) 존재 확인.

# 2. 모든 dataset에서 column 동일한지 확인

모든 OTT 데이터셋의 컬럼이 동일한지 검증 완료. ✅

In [14]:
# 각 데이터프레임의 컬럼 확인
print("🎬 Disney+ 컬럼:", disney_df.columns)
print("🎬 Amazon Prime 컬럼:", amazon_df.columns)
print("🎬 Netflix 컬럼:", netflix_df.columns)

# 모든 데이터셋에서 컬럼이 동일한지 확인
print("✅ 모든 데이터셋의 컬럼 동일 여부:",
      disney_df.columns.equals(amazon_df.columns) and disney_df.columns.equals(netflix_df.columns))

🎬 Disney+ 컬럼: Index(['show_id', 'type', 'title', 'director', 'cast', 'release_year',
       'date_added', 'listed_in', 'duration', 'platform'],
      dtype='object')
🎬 Amazon Prime 컬럼: Index(['show_id', 'type', 'title', 'director', 'cast', 'release_year',
       'date_added', 'listed_in', 'duration', 'platform'],
      dtype='object')
🎬 Netflix 컬럼: Index(['show_id', 'type', 'title', 'director', 'cast', 'release_year',
       'date_added', 'listed_in', 'duration', 'platform'],
      dtype='object')
✅ 모든 데이터셋의 컬럼 동일 여부: True


# 결측치 확인

In [15]:
# 데이터프레임별 결측치 개수 확인
print("🔍 Disney+ 결측치 개수:\n", disney_df.isnull().sum())
print("\n🔍 Amazon Prime 결측치 개수:\n", amazon_df.isnull().sum())
print("\n🔍 Netflix 결측치 개수:\n", netflix_df.isnull().sum())

# 전체 데이터 중 결측치 비율 확인
print("\n🔎 Disney+ 결측치 비율:\n", (disney_df.isnull().sum() / len(disney_df)) * 100)
print("\n🔎 Amazon Prime 결측치 비율:\n", (amazon_df.isnull().sum() / len(amazon_df)) * 100)
print("\n🔎 Netflix 결측치 비율:\n", (netflix_df.isnull().sum() / len(netflix_df)) * 100)

🔍 Disney+ 결측치 개수:
 show_id           0
type              0
title             0
director        473
cast            190
release_year      0
date_added        3
listed_in         0
duration          0
platform          0
dtype: int64

🔍 Amazon Prime 결측치 개수:
 show_id            0
type               0
title              0
director        2083
cast            1233
release_year       0
date_added      9513
listed_in          0
duration           0
platform           0
dtype: int64

🔍 Netflix 결측치 개수:
 show_id            0
type               0
title              0
director        2634
cast             825
release_year       0
date_added        10
listed_in          0
duration           3
platform           0
dtype: int64

🔎 Disney+ 결측치 비율:
 show_id          0.000000
type             0.000000
title            0.000000
director        32.620690
cast            13.103448
release_year     0.000000
date_added       0.206897
listed_in        0.000000
duration         0.000000
platform         0.0000

# 중복data 확인

In [16]:
# 데이터프레임별 중복 데이터 개수 확인
print("📌 Disney+ 중복 데이터 개수:", disney_df.duplicated().sum())
print("📌 Amazon Prime 중복 데이터 개수:", amazon_df.duplicated().sum())
print("📌 Netflix 중복 데이터 개수:", netflix_df.duplicated().sum())

📌 Disney+ 중복 데이터 개수: 0
📌 Amazon Prime 중복 데이터 개수: 0
📌 Netflix 중복 데이터 개수: 0


# 결측치 처리 및 data 병합

In [17]:
# 1️⃣ 결측치 처리
for df in [disney_df, amazon_df, netflix_df]:
    df = df.copy()  # 경고 방지를 위해 복사본 사용

    # 'date_added' 컬럼이 존재하면 삭제
    if 'date_added' in df.columns:
        df = df.drop(columns=['date_added'])

    # 'director'와 'cast'의 결측치를 'Unknown'으로 채우기
    df['director'] = df['director'].fillna('Unknown')
    df['cast'] = df['cast'].fillna('Unknown')

    # 'duration'의 결측치가 3개뿐이므로 해당 행 제거
    df = df.dropna(subset=['duration'])

# 2️⃣ 데이터 병합
combined_df = pd.concat([disney_df, amazon_df, netflix_df], ignore_index=True)

# 확인
print("✅ 최종 데이터 크기:", combined_df.shape)
print(combined_df.head())

✅ 최종 데이터 크기: (19925, 10)
  show_id     type                                             title  \
0      s1    Movie  Duck the Halls: A Mickey Mouse Christmas Special   
1      s2    Movie                            Ernest Saves Christmas   
2      s3    Movie                      Ice Age: A Mammoth Christmas   
3      s4    Movie                        The Queen Family Singalong   
4      s5  TV Show                             The Beatles: Get Back   

                            director  \
0  Alonso Ramirez Ramos, Dave Wasson   
1                        John Cherry   
2                       Karen Disher   
3                    Hamish Hamilton   
4                                NaN   

                                                cast  release_year  \
0  Chris Diamantopoulos, Tony Anselmo, Tress MacN...          2016   
1           Jim Varney, Noelle Parker, Douglas Seale          1988   
2  Raymond Albert Romano, John Leguizamo, Denis L...          2011   
3  Darren Criss, Adam

# 중복되는 영화/TV프로그램 제거

In [18]:
# 중복된 영화/TV 프로그램을 하나만 남기고 삭제
# 우선순위 설정: Netflix > Amazon Prime > Disney+
platform_order = ['Netflix', 'Amazon', 'Disney']

# 중복된 title에 대해 platform 기준으로 우선순위를 적용하여 남길 행을 결정
df_sorted = df.sort_values(by=['title', 'platform'], ascending=[True, True])
df_sorted['platform'] = pd.Categorical(df_sorted['platform'], categories=platform_order, ordered=True)

# 중복된 title을 가진 행에서 platform 순서대로 중복 제거
df_deduplicated = df_sorted.drop_duplicates(subset=['title'], keep='first')

# 결과 확인
print(f'중복 제거 후 데이터 크기: {df_deduplicated.shape}')

중복 제거 후 데이터 크기: (8804, 9)


# 장르 파악 (listed_in)

상위 20개 파악

In [19]:
# 1단계: listed_in 컬럼에서 장르 목록을 추출하여 장르의 출현 빈도 계산하기
# 각 영화/TV 프로그램의 listed_in이 여러 개의 장르를 포함할 수 있으므로, 먼저 이를 분리하고 나열된 장르들을 확인

# 'listed_in' 컬럼에서 장르를 분리하여 나열
all_genres = df_deduplicated['listed_in'].str.split(',').explode().str.strip()

# 장르별 출현 빈도 계산
genre_counts = all_genres.value_counts()

# 결과 확인
print(genre_counts.head(20))  # 상위 20개의 장르를 출력

listed_in
International Movies        2752
Dramas                      2427
Comedies                    1674
International TV Shows      1351
Documentaries                869
Action & Adventure           859
TV Dramas                    763
Independent Movies           756
Children & Family Movies     641
Romantic Movies              616
TV Comedies                  581
Thrillers                    577
Crime TV Shows               470
Kids' TV                     451
Docuseries                   395
Music & Musicals             375
Romantic TV Shows            370
Horror Movies                357
Stand-Up Comedy              343
Reality TV                   255
Name: count, dtype: int64


전체 장르 파악

In [20]:
# 1. listed_in 컬럼에 등장하는 고유한 값들을 확인
unique_genres = df_deduplicated['listed_in'].unique()

# 2. 고유한 장르 목록 출력
print(unique_genres)

['Horror Movies, International Movies, Thrillers'
 'Documentaries, International Movies'
 'Dramas, International Movies, Romantic Movies'
 'Comedies, Romantic Movies' 'Documentaries, Sports Movies'
 'Comedies, Dramas, International Movies' 'TV Comedies' 'Comedies'
 'Dramas, International Movies' 'Sports Movies' 'Documentaries'
 'Reality TV' 'Dramas, Romantic Movies'
 'Dramas, Romantic Movies, Sports Movies'
 'Comedies, International Movies, Romantic Movies' 'Action & Adventure'
 'International TV Shows, Romantic TV Shows, TV Dramas'
 'Docuseries, Science & Nature TV'
 'Dramas, International Movies, Sports Movies' 'Movies'
 'British TV Shows, International TV Shows, Reality TV'
 'Dramas, Independent Movies, Sports Movies'
 'International TV Shows, Korean TV Shows, Romantic TV Shows'
 'Horror Movies, International Movies'
 'Horror Movies, Independent Movies, Thrillers'
 'Crime TV Shows, TV Dramas, TV Mysteries' 'Crime TV Shows, Docuseries'
 'Horror Movies, Thrillers' 'Action & Adventure,

# 장르 개수 파악 및 대분류 카테고리화

In [21]:
import re

# 작은따옴표 제거 후, ','로 세부 장르 분리하기
cleaned_genres = []

for genre in unique_genres:
    # 작은따옴표 제거
    genre = genre.replace("'", "")
    # ','로 세부 장르 분리
    cleaned_genres.extend(genre.split(', '))

# 중복 제거 후 유니크한 세부 장르 세기
unique_cleaned_genres = set(cleaned_genres)
num_unique_genres = len(unique_cleaned_genres)


print(f"유니크한 세부 장르의 개수: {num_unique_genres}")
print("유니크한 세부 장르 목록:")
for genre in unique_cleaned_genres:
    print(genre)

유니크한 세부 장르의 개수: 42
유니크한 세부 장르 목록:
Cult Movies
Classic Movies
Stand-Up Comedy
Dramas
Kids TV
Sci-Fi & Fantasy
Teen TV Shows
British TV Shows
LGBTQ Movies
TV Horror
Action & Adventure
Horror Movies
TV Action & Adventure
Classic & Cult TV
Crime TV Shows
Music & Musicals
TV Dramas
Thrillers
Faith & Spirituality
Spanish-Language TV Shows
International TV Shows
International Movies
Children & Family Movies
TV Shows
Anime Series
Comedies
Korean TV Shows
TV Comedies
Romantic TV Shows
Science & Nature TV
Sports Movies
Stand-Up Comedy & Talk Shows
Documentaries
Docuseries
TV Mysteries
Independent Movies
TV Thrillers
TV Sci-Fi & Fantasy
Movies
Anime Features
Romantic Movies
Reality TV


# 10개 대분류 카테고리와 해당 세부 장르 매핑

In [22]:
# 10개 대분류 카테고리와 해당 세부 장르 매핑
genre_mapping = {
    'Comedy': [
        'Stand-Up Comedy & Talk Shows', 'TV Comedies', 'Comedies', 'Stand-Up Comedy'
    ],
    'Romance': [
        'Romantic TV Shows', 'Romantic Movies'
    ],
    'Movies': [
        'Classic Movies', 'Horror Movies', 'Independent Movies', 'Cult Movies', 'Movies'
    ],
    'Documentary': [
        'Docuseries', 'Documentaries'
    ],
    'Animation': [
        'Anime Features', 'Anime Series'
    ],
    'Sports': [
        'Sports Movies', 'Reality TV'
    ],
    'Drama': [
        'TV Dramas', 'Dramas', 'TV Horror', 'Sci-Fi & Fantasy', 'TV Action & Adventure',
        'TV Sci-Fi & Fantasy', 'TV Thrillers', 'TV Mysteries', 'Crime TV Shows'
    ],
    'Religion': [
        'Faith & Spirituality'
    ],
    'Family': [
        'Children & Family Movies', 'Kids TV', 'TV Shows'
    ],
    'Sci-Fi': [
        'Sci-Fi & Fantasy', 'TV Sci-Fi & Fantasy', 'TV Action & Adventure',
        'Thrillers', 'TV Action & Adventure'
    ]
}

# 결과 출력 (딕셔너리 구조 확인)
for category, genres in genre_mapping.items():
    print(f"{category}: {genres}")

Comedy: ['Stand-Up Comedy & Talk Shows', 'TV Comedies', 'Comedies', 'Stand-Up Comedy']
Romance: ['Romantic TV Shows', 'Romantic Movies']
Movies: ['Classic Movies', 'Horror Movies', 'Independent Movies', 'Cult Movies', 'Movies']
Documentary: ['Docuseries', 'Documentaries']
Animation: ['Anime Features', 'Anime Series']
Sports: ['Sports Movies', 'Reality TV']
Drama: ['TV Dramas', 'Dramas', 'TV Horror', 'Sci-Fi & Fantasy', 'TV Action & Adventure', 'TV Sci-Fi & Fantasy', 'TV Thrillers', 'TV Mysteries', 'Crime TV Shows']
Religion: ['Faith & Spirituality']
Family: ['Children & Family Movies', 'Kids TV', 'TV Shows']
Sci-Fi: ['Sci-Fi & Fantasy', 'TV Sci-Fi & Fantasy', 'TV Action & Adventure', 'Thrillers', 'TV Action & Adventure']


# 한 영화/TV프로그램 당 한 장르 배정하기

listed_in colunm외에도

category와 category_list	column과 top_category column 활용

In [23]:
# 4️⃣ 세부 장르를 대분류 카테고리로 변환하는 함수
def map_genres_to_categories(listed_in, genre_mapping):
    # 작은따옴표 제거
    listed_in = listed_in.replace("'", "")

    # 결과로 대분류 카테고리 제목만 담을 리스트
    categories = []

    # 세부 장르들을 대분류 카테고리로 매핑
    for category, genres in genre_mapping.items():
        for genre in genres:
            if genre in listed_in:
                categories.append(category)
                break  # 하나의 대분류 카테고리만 추가 (중복 방지)

    # 대분류 카테고리들이 없으면 빈 문자열 반환
    return ', '.join(categories) if categories else ''

# 5️⃣ 'listed_in' 컬럼을 대분류 카테고리로 변환
df_deduplicated['category'] = df_deduplicated['listed_in'].apply(lambda x: map_genres_to_categories(x, genre_mapping))

# 6️⃣ 결과 확인
print(df_deduplicated[['title', 'platform', 'listed_in', 'category']])

from IPython.display import display

display(df_deduplicated.head(20))

                              title platform  \
2036                         #Alive  Netflix   
2304  #AnneFrank - Parallel Stories  Netflix   
2481              #FriendButMarried  Netflix   
2324            #FriendButMarried 2  Netflix   
5973                          #Roxy  Netflix   
...                             ...      ...   
6177                       忍者ハットリくん  Netflix   
4914                           海的儿子  Netflix   
7101                           마녀사냥  Netflix   
5022                        반드시 잡는다  Netflix   
7108            최강전사 미니특공대 : 영웅의 탄생  Netflix   

                                              listed_in  \
2036     Horror Movies, International Movies, Thrillers   
2304                Documentaries, International Movies   
2481      Dramas, International Movies, Romantic Movies   
2324      Dramas, International Movies, Romantic Movies   
5973                          Comedies, Romantic Movies   
...                                                 ...   
6177      

show_id     type                          title  \
2036   s2037    Movie                         #Alive   
2304   s2305    Movie  #AnneFrank - Parallel Stories   
2481   s2482    Movie              #FriendButMarried   
2324   s2325    Movie            #FriendButMarried 2   
5973   s5974    Movie                          #Roxy   
5695   s5696    Movie                      #Rucker50   
654     s655    Movie                        #Selfie   
655     s656    Movie                     #Selfie 69   
2666   s2667  TV Show                       #blackAF   
5972   s5973    Movie               #cats_the_mewvie   
5277   s5278    Movie                   #realityhigh   
306     s307    Movie                            '76   
5967   s5968    Movie                            '89   
5971   s5972    Movie                       (T)ERROR   
2140   s2141  TV Show                       (Un)Well   
5979   s5980    Movie               1 Chance 2 Dance   
5980   s5981    Movie                  1 Mile to You   
3392   s3393    Movie            10 Days in Sun City   
5387   s5388    Movie                 10 jours en or   
5981   s5982    Movie                    10,000 B.C.   

                                    director  \
2036                                  Cho Il   
2304             Sabina Fedeli, Anna Migotto   
2481                           Rako Prijanto   
2324                           Rako Prijanto   
5973                         Michael Kennedy   
5695                   Robert McCullough Jr.   
654                           Cristina Jacob   
655                           Cristina Jacob   
2666                                 Unknown   
5972                        Michael Margolis   
5277                        Fernando Lebrija   
306                               Izu Ojukwu   
5967                                 Unknown   
5971  Lyric R. Cabral, David Felix Sutcliffe   
2140                                 Unknown   
5979                              Adam Deyoe   
5980                             Leif Tilden   
3392                               Adze Ugah   
5387                       Nicolas Brossette   
5981                         Roland Emmerich   

                                                   cast  release_year  \
2036                           Yoo Ah-in, Park Shin-hye          2020   
2304                        Helen Mirren, Gengher Gatti          2019   
2481  Adipati Dolken, Vanesha Prescilla, Rendi Jhon,...          2018   
2324  Adipati Dolken, Mawar de Jongh, Sari Nila, Von...          2020   
5973  Jake Short, Sarah Fisher, Booboo Stewart, Dann...          2018   
5695                                            Unknown          2016   
654   Flavia Hojda, Crina Semciuc, Olimpia Melinte, ...          2014   
655   Maia Morgenstern, Olimpia Melinte, Crina Semci...          2016   
2666  Kenya Barris, Rashida Jones, Iman Benson, Genn...          2020   
5972                                            Unknown          2020   
5277  Nesta Cooper, Kate Walsh, John Michael Higgins...          2017   
306   Ramsey Nouah, Rita Dominic, Chidi Mokeme, Ibin...          2016   
5967                 Lee Dixon, Ian Wright, Paul Merson          2017   
5971                                            Unknown          2015   
2140                                            Unknown          2020   
5979  Lexi Giovagnoli, Justin Ray, Rae Latt, Poonam ...          2014   
5980  Billy Crudup, Graham Rogers, Liana Liberato, S...          2017   
3392  Ayo Makun, Adesua Etomi, Richard Mofe-Damijo, ...          2017   
5387  Franck Dubosc, Claude Rich, Marie Kremer, Math...          2012   
5981  Steven Strait, Camilla Belle, Cliff Curtis, Jo...          2008   

                                            listed_in  duration platform  \
2036   Horror Movies, International Movies, Thrillers    99 min  Netflix   
2304              Documentaries, International Movies    95 min  Netflix   
2481    Dramas, International Movies, Romantic Movies   1

In [24]:
import pandas as pd
from collections import Counter

# df_deduplicated를 기반으로 작업
# 각 작품의 category를 쉼표로 분할하여 리스트로 변환
df_deduplicated["category_list"] = df_deduplicated["listed_in"].apply(lambda x: x.split(", "))

# 모든 category를 모아서 빈도수 계산
all_categories = [cat for sublist in df_deduplicated["category_list"] for cat in sublist]
category_counts = Counter(all_categories)

# 카테고리를 빈도수 기준으로 정렬
top_categories = sorted(category_counts.items(), key=lambda x: x[1], reverse=True)
print("카테고리 빈도수 순위:", top_categories)

# 각 작품에 대해 가장 우선순위가 높은 대분류 배정
def get_top_category(categories):
    sorted_categories = sorted(categories, key=lambda x: category_counts[x], reverse=True)
    return sorted_categories[0] if sorted_categories else None

df_deduplicated["top_category"] = df_deduplicated["category_list"].apply(get_top_category)

# 결과 출력
print(df_deduplicated[["title", "top_category"]])

display(df_deduplicated.head(20))

카테고리 빈도수 순위: [('International Movies', 2752), ('Dramas', 2427), ('Comedies', 1674), ('International TV Shows', 1351), ('Documentaries', 869), ('Action & Adventure', 859), ('TV Dramas', 763), ('Independent Movies', 756), ('Children & Family Movies', 641), ('Romantic Movies', 616), ('TV Comedies', 581), ('Thrillers', 577), ('Crime TV Shows', 470), ("Kids' TV", 451), ('Docuseries', 395), ('Music & Musicals', 375), ('Romantic TV Shows', 370), ('Horror Movies', 357), ('Stand-Up Comedy', 343), ('Reality TV', 255), ('British TV Shows', 253), ('Sci-Fi & Fantasy', 243), ('Sports Movies', 219), ('Anime Series', 176), ('Spanish-Language TV Shows', 174), ('TV Action & Adventure', 168), ('Korean TV Shows', 151), ('Classic Movies', 116), ('LGBTQ Movies', 102), ('TV Mysteries', 98), ('Science & Nature TV', 92), ('TV Sci-Fi & Fantasy', 84), ('TV Horror', 75), ('Cult Movies', 71), ('Anime Features', 71), ('Teen TV Shows', 69), ('Faith & Spirituality', 65), ('TV Thrillers', 57), ('Stand-Up Comedy & Talk

show_id     type                          title  \
2036   s2037    Movie                         #Alive   
2304   s2305    Movie  #AnneFrank - Parallel Stories   
2481   s2482    Movie              #FriendButMarried   
2324   s2325    Movie            #FriendButMarried 2   
5973   s5974    Movie                          #Roxy   
5695   s5696    Movie                      #Rucker50   
654     s655    Movie                        #Selfie   
655     s656    Movie                     #Selfie 69   
2666   s2667  TV Show                       #blackAF   
5972   s5973    Movie               #cats_the_mewvie   
5277   s5278    Movie                   #realityhigh   
306     s307    Movie                            '76   
5967   s5968    Movie                            '89   
5971   s5972    Movie                       (T)ERROR   
2140   s2141  TV Show                       (Un)Well   
5979   s5980    Movie               1 Chance 2 Dance   
5980   s5981    Movie                  1 Mile to You   
3392   s3393    Movie            10 Days in Sun City   
5387   s5388    Movie                 10 jours en or   
5981   s5982    Movie                    10,000 B.C.   

                                    director  \
2036                                  Cho Il   
2304             Sabina Fedeli, Anna Migotto   
2481                           Rako Prijanto   
2324                           Rako Prijanto   
5973                         Michael Kennedy   
5695                   Robert McCullough Jr.   
654                           Cristina Jacob   
655                           Cristina Jacob   
2666                                 Unknown   
5972                        Michael Margolis   
5277                        Fernando Lebrija   
306                               Izu Ojukwu   
5967                                 Unknown   
5971  Lyric R. Cabral, David Felix Sutcliffe   
2140                                 Unknown   
5979                              Adam Deyoe   
5980                             Leif Tilden   
3392                               Adze Ugah   
5387                       Nicolas Brossette   
5981                         Roland Emmerich   

                                                   cast  release_year  \
2036                           Yoo Ah-in, Park Shin-hye          2020   
2304                        Helen Mirren, Gengher Gatti          2019   
2481  Adipati Dolken, Vanesha Prescilla, Rendi Jhon,...          2018   
2324  Adipati Dolken, Mawar de Jongh, Sari Nila, Von...          2020   
5973  Jake Short, Sarah Fisher, Booboo Stewart, Dann...          2018   
5695                                            Unknown          2016   
654   Flavia Hojda, Crina Semciuc, Olimpia Melinte, ...          2014   
655   Maia Morgenstern, Olimpia Melinte, Crina Semci...          2016   
2666  Kenya Barris, Rashida Jones, Iman Benson, Genn...          2020   
5972                                            Unknown          2020   
5277  Nesta Cooper, Kate Walsh, John Michael Higgins...          2017   
306   Ramsey Nouah, Rita Dominic, Chidi Mokeme, Ibin...          2016   
5967                 Lee Dixon, Ian Wright, Paul Merson          2017   
5971                                            Unknown          2015   
2140                                            Unknown          2020   
5979  Lexi Giovagnoli, Justin Ray, Rae Latt, Poonam ...          2014   
5980  Billy Crudup, Graham Rogers, Liana Liberato, S...          2017   
3392  Ayo Makun, Adesua Etomi, Richard Mofe-Damijo, ...          2017   
5387  Franck Dubosc, Claude Rich, Marie Kremer, Math...          2012   
5981  Steven Strait, Camilla Belle, Cliff Curtis, Jo...          2008   

                                            listed_in  duration platform  \
2036   Horror Movies, International Movies, Thrillers    99 min  Netflix   
2304              Documentaries, International Movies    95 min  Netflix   
2481    Dramas, International Movies, Romantic Movies   1

# 이 아래는0218 0219 과정

# 지금은 0220 오늘 과정 우선으로 처음부터 진행 중

#column 수정

- 세 개의 데이터프레임(netflix_df, amazon_df, disney_df)의 컬럼 이름을 소문자로 변경하고, 공통 컬럼만 남기기

- 각 데이터프레임에 'platform' 컬럼을 추가하여 데이터 출처를 'Netflix', 'Amazon', 'Disney'로 구분

In [ ]:
# 컬럼 이름을 소문자로 통일
netflix_df.columns = netflix_df.columns.str.lower()
amazon_df.columns = amazon_df.columns.str.lower()
disney_df.columns = disney_df.columns.str.lower()

# 공통 컬럼만 추출 (필요한 컬럼만 남기는 것으로 활용해도 OK)
common_cols = ['show_id', 'type', 'title', 'director', 'cast', 'release_year', 'date_added','listed_in','duration']
netflix_df = netflix_df[common_cols]
amazon_df = amazon_df[common_cols]
disney_df = disney_df[common_cols]

# 출처 플랫폼 컬럼 추가
netflix_df['platform'] = 'Netflix'
amazon_df['platform'] = 'Amazon'
disney_df['platform'] = 'Disney'

amazon_df.head(20)

#3사 Data 합치기

- 세 data set 모두 column 12개로 내용도 동일한 상황

- 세 개의 데이터프레임(netflix_df, amazon_df, disney_df)을 하나로 합쳐 df 생성

- pd.concat을 이용하여 데이터프레임들을 연결하고, join='inner' 옵션으로 공통 컬럼만 유지

- df['platform'].sample(frac=1)을 통해 컬럼 값을 랜덤하게 섞고, reset_index를 통해 인덱스를 재설정

In [ ]:
# 공통 컬럼만 유지하고 행 추가 (inner join)
df = pd.concat([netflix_df, amazon_df, disney_df], join='inner', ignore_index=True)

# 결과 확인
df.head(10)

#cast column 분리

- 'cast' 컬럼의 값을 ','로 분리하여 배우 이름을 리스트 형태로 'actor' 컬럼에 저장 (apply와 lambda 함수)

- 'actor' 컬럼을 기준으로 각 배우가 별도의 행을 갖도록 진행

- explode를 사용하여 'actor' 컬럼의 리스트 값을 여러 행으로 확장했습니다.

- 'actor' 컬럼의 값에서 str.strip() 사용하여 배우 이름의 앞뒤 공백을 제거

In [ ]:
# 'cast' 컬럼을 ',' 기준으로 분리해서 리스트로 만들기
df['actor'] = df['cast'].apply(lambda x: str(x).split(','))

# 'actor' 컬럼에서 대괄호 제거
df['actor'] = df['actor'].apply(lambda x: [item.replace('[', '').replace(']', '').strip() for item in x])

# 같은 영화에 출연한 배우(cast 컬럼의 리스트)를 여러 행으로 확장
df_exploded = df.explode('actor')

# 배우 이름 앞뒤 공백 제거
df_exploded['actor'] = df_exploded['actor'].str.strip()

# 결과 확인 (타이틀과 배우 컬럼만 보기)
df_exploded[['title', 'actor']].head(50)

#listed_in column 분리

- 'listed_in' 컬럼의 값을 쉼표(,)로 분리하여 장르 리스트를 생성하고, 'genre_list' 컬럼에 저장

- str.split()과 explode()를 이용하여 장르 정보를 분리하고 데이터프레임을 확장

- 'title' 컬럼을 유지하면서 'genre_list' 컬럼을 여러 행으로 확장하여 작품과 장르 정보를 연결

- explode()를 사용하여 각 장르가 별도의 행을 갖도록 데이터프레임을 변형

In [ ]:
# 'listed_in' 컬럼의 데이터 타입을 문자열로 변환하고 NaN 값을 빈 문자열로 대체
df['listed_in'] = df['listed_in'].astype(str).fillna('')

# 'listed_in' 컬럼의 값을 쉼표(`, `)로 분리하여 장르 리스트 생성
df['genre_list'] = df['listed_in'].apply(lambda x: x.split(', ')) # explode 이전에 str.split 적용

# title을 유지하면서 genre_list를 여러 행으로 확장
df_exploded_genre = df[['title', 'genre_list']].explode('genre_list')

df_exploded_genre['genre_list'] = df_exploded_genre['genre_list'].str.replace(r"[\[\]]", "", regex=True)

# 결과 확인
df_exploded_genre.head(10)

#listed_in column 내 등장하는 장르 확인

총 102개

In [ ]:
unique_genres = df_exploded_genre['genre_list'].nunique()
print(f"현재 데이터셋 내 고유한 장르 개수: {unique_genres}")

In [ ]:
genre_counts = df_exploded_genre['genre_list'].value_counts()
print(genre_counts.head(51))
print(genre_counts.tail(51))

#Unscripted 라는 워딩 -> 오류 아닌지 확인

In [ ]:
# 'Unscripted'가 실제 데이터에 존재하는지 확인
print(df['listed_in'].unique())  # 고유한 장르 리스트 출력

# 'Unscripted'가 있는 행 출력
print(df[df['listed_in'].str.contains('Unscripted', na=False)])

#cast column과 listed_in column 따로 빼기 + 결측치 채우기

- cast column에는 3가지 data set 모두 약 10% 정도의 null값 존재

- listed_in column에는 null값 없음

- 별점에 대한 imdb data set은 나중에 다시 불러와서 비교분석할 예정

In [ ]:
df = df.copy()  # 혹시 모를 경고 방지를 위해 복사본 생성
df['cast'] = df['cast'].fillna('Unknown')

df['cast'].isnull().sum()

In [ ]:
import pandas as pd
import re

# 텍스트 정제 함수
def clean_text_list(text_list):
    return ["".join(re.findall(r"[A-Za-z0-9\s]", text)).strip() for text in text_list if text]

# 텍스트 정제
if not isinstance(df['title'].iloc[0], list):
    df["title"] = df["title"].fillna("").apply(lambda x: x.split(", ") if x else [])
if not isinstance(df['listed_in'].iloc[0], list):
    df["listed_in"] = df["listed_in"].fillna("").apply(lambda x: x.split(", ") if x else [])
if not isinstance(df['platform'].iloc[0], list):
    df["platform"] = df["platform"].fillna("").apply(lambda x: x.split(", ") if x else [])

if not re.fullmatch(r"[A-Za-z0-9\s]+", str(df['title'].iloc[0][0] if df['title'].iloc[0] else '')):
    df["title"] = df["title"].apply(clean_text_list)
if not re.fullmatch(r"[A-Za-z0-9\s]+", str(df['listed_in'].iloc[0][0] if df['listed_in'].iloc[0] else '')):
    df["listed_in"] = df["listed_in"].apply(clean_text_list)
if not re.fullmatch(r"[A-Za-z0-9\s]+", str(df['platform'].iloc[0][0] if df['platform'].iloc[0] else '')):
    df["platform"] = df["platform"].apply(clean_text_list)
if not re.fullmatch(r"[A-Za-z0-9\s]+", str(df['cast'].iloc[0][0] if df['cast'].iloc[0] else '')):
    df["cast"] = df["cast"].apply(clean_text_list)

# 결과 확인
df_cleaned.head()